Fix unnecessary exception conversion by `MaskSensitiveExceptions`

jeeminso opened the pull request:
## Summary of the changes / Why this improves CrateDB
Fixes https://github.com/crate/crate/issues/13561

`ColumnUnknownException` implements `TableScopedException` where `TableScopedException` enables privilege checks on the associated `RelationName`. However, table functions such as `empty_row()` are not tied to any schema/relations and also privilege checks are not required which make the current `ColumnUnknownException` not suitable (for cases like `select unknown_column from abs(1)`). 

The approach taken is 1) to refactor the exceptions package grouped by the scope (table, schema, cluster, unscoped, tablefuntion), and 2) create another `ColumnUnknownException` in `tablefunction` package to handle privilege checks on table functions separate from regular tables. 



## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

matriv commented on server/src/main/java/io/crate/auth/AccessControlImpl.java:
```diff
@@ -906,7 +907,7 @@ protected Void visitSchemaScopeException(SchemaScopeException e, User context) {
 
         @Override
-        protected Void visitClusterScopeException(ClusterScopeException e, User context) {
+        public Void visitClusterScopeException(ClusterScopeException e, User context) {
```
why do you need public here?

matriv commented on server/src/main/java/io/crate/exceptions/CrateExceptionVisitor.java:
```diff
@@ -41,11 +41,15 @@ protected R visitSchemaScopeException(SchemaScopeException e, C context) {
     }
 
-    protected R visitClusterScopeException(ClusterScopeException e, C context) {
+    public R visitClusterScopeException(ClusterScopeException e, C context) {
```
and here

matriv commented on server/src/main/java/io/crate/exceptions/scoped/tablefunction/ColumnUnknownException.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static ColumnUnknownException columnUnknownFromTableFunctionException(
+        String columnName, RelationName relationName
+    ) {
```
identation seems wrong, shouldn't this be in the prev line?

matriv commented on server/src/main/java/io/crate/exceptions/scoped/tablefunction/ColumnUnknownException.java:
```diff
@@ -0,0 +1,52 @@
+import io.crate.exceptions.TableFunctionScopeException;
+import io.crate.metadata.RelationName;
+
+public class ColumnUnknownException extends RuntimeException implements ResourceUnknownException, TableFunctionScopeException {
```
I would rather rename the exception to `ColumnFromTableFunctionUnknownException`
and a void the static method below.

matriv reviewed (commented):
Thank you for the work on that!
I've left some comments, but also overall I don't like this distinction of exceptions based on a different package. I'd rather keep the same package as it was and rename the exceptions to match their context. It can be confusing to have exceptions with the same name, especially used in the same class where you have to use fully qualified name  and not imports.

mfussenegger commented on server/src/main/java/io/crate/analyze/relations/FullQualifiedNameFieldProvider.java:
```diff
@@ -156,7 +158,11 @@ public Symbol resolveField(QualifiedName qualifiedName, @Nullable List<String> p
+        var source = sources.entrySet().iterator().next();
+        RelationName relationName = source.getKey();
+        if (source.getValue() instanceof TableFunctionRelation) {
+            throw columnUnknownFromTableFunctionException(columnIdent.sqlFqn(), relationName);
```
Maybe we need to rethink the scoping system,  but we shouldn't need different kind of unknown column exceptions and components raising them shouldn't need to know which one of them to raise

jeeminso commented:
Thanks for reviewing @matriv and @mfussenegger. Since this seems like not the way to go, instead of addressing the reviews I have come up with another approach.

I had an impression that only `FieldProvider.resolveField` should throw `ColumnUnknownExceptions` not `AnalyzedRelation.getField`. But in this case allowing `TableFunctionRelation.getField` to throw `ColumnUnknownException` makes things easier. Please check https://github.com/crate/crate/pull/13716.
